lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp

```diff
@@ -252,6 +273,219 @@ void FIRRTLTypesLowering::visitDecl(InstanceOp op) {
+          FIRRTLType theType = a.type.cast<FIRRTLType>();
+          if (isFlipped)
+            theType = FlipType::get(theType);
+
+          auto subfield = builder->create<SubfieldOp>(
+              theType, newMem.getResult(i), elt.name);
+          setBundleLowering(op.getResult(i), elt.name.str() + a.suffix,
+                            subfield);
+        }
+      }
+    }
+  }
+
+  // llvm::errs() << "Memory will be size: " << width << "\n";
+
+  // auto newName = op.name().getValue().str() + "_flattened";
+
+  // auto flatDataType = UIntType::get(builder->getContext(), width);
+
+  // SmallVector<Type, 8> resultTypes;
+  // for (size_t i = 0, e = op.getNumResults(); i != e; ++i) {
+  //   SmallVector<BundleType::BundleElement, 5> resultFields;
+  //   for (auto elt :
+  //        op.getResult(i).getType().cast<BundleType>().getElements()) {
+  //     if (elt.name == "data") {
+  //       resultFields.push_back({elt.name, flatDataType});
+  //       continue;
+  //     }
+  //     resultFields.push_back(elt);
+  //   }
+  //   resultTypes.push_back(BundleType::get(resultFields,
+  //   builder->getContext()));
+  // }
+
+  // auto newMem = builder->create<MemOp>(
+  //     resultTypes, op.readLatency(), op.writeLatency(), op.depth(), op.ruw(),
+  //     op.portNames(), builder->getStringAttr(newName));
+
+  // llvm::errs() << newMem << "\n";
+
+  // // Create bit accessors for each bit.
+  // for (size_t i = 0, e = op.getNumResults(); i != e; ++i) {
+  //   auto port = newMem.getResult(i);
+  //   for (auto a : fieldTypes) {
+  //     auto b = a.type.getBitWidthOrSentinel();
+  //     llvm::errs() << "  - suffix: " << a.suffix << "\n    type: " << a.type
+  //                  << "\n  - position: " << width << "\n    width: " << b
+  //                  << "\n";
+  //     auto subfield = builder->create<SubfieldOp>(flatDataType, port,
+  //     "data"); auto bitExtract =
+  //         builder->create<BitsPrimOp>(subfield, width - 1, width - b);
+  //     setBundleLowering(op.getResult(i), "data" + a.suffix, bitExtract);
+  //     width -= b;
+  //   }
+  //   for (auto a : port.getType().cast<BundleType>().getElements()) {
+  //     if (a.name == "data")
+  //       continue;
+  //     auto subfield = builder->create<SubfieldOp>(a.type, port, a.name);
+  //     setBundleLowering(op.getResult(i), a.name, subfield);
+  //   }
+  // }
+
+  dumpFrom(op->getParentOfType<firrtl::FModuleOp>());
+
+  // for (auto a : fieldTypes) {
+  //   llvm::errs() << "  - suffix: " << a.suffix << "\n";
+  //   llvm::errs() << "    type: " << a.type << "\n";
+
+  //   auto newName = op.name().getValue().str() + a.suffix;
+
+  //   for (size_t i = 0, e = op.getNumResults(); i != e; ++i) {
+  //     for (size_t j = 0, e = newMem.getNumResults(); j != e; ++j) {
+  //       // llvm::errs() << "i: " << i << ", j: " << j << "\n";
+  //       // llvm::errs() << "setBundleLowering:\n"
+  //       //              << "  old: " << op.getResult(i) << "\n"
```

Comment:
same :-).  I don't think this patch should add new functionality to LowerTypes.  This should be split out to a separate patch

Reply:
To clarify, you're only talking about this commented out code not being included here, yes? Or, are you asking about splitting the patch into: (1) returning multiple results and (2) adding support for memory lowering in `LowerTypes`?

(Assuming the former is what you mean...) Can do. Returning multiple results seemed to blow up lower types in weird ways hence this patch *also* updating lower types. The commented out code was the first effort of lowering by flattening the memory type discussed in the [second enumeration in this comment](https://github.com/llvm/circt/issues/478#issuecomment-762491758).  (The SFC also supports this with`--custom-transforms firrtl.transforms.SimplifyMems` option.) *However, that can't universally be used if the memories have write ports with masks that are not tied to their enable lines.* I agree that that adding memory type flattening ("simplification") support would make a nice follow-on patch.

I'll move this logic to a separate branch as opposed to deleting it.

Reply:
This commented out "simplify mems" logic is now on branch: [`dev/seldridge/simplify-memories` branch](https://github.com/llvm/circt/tree/dev/seldridge/simplify-memories).